File: js_package/everyai-core/src/governor.js
/**
 * Helper to pause execution for a given number of milliseconds.
 * @param {number} ms 
 * @returns {Promise<void>}
 */
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Locally tracks request history to govern and throttle API request rates.
 */
export class RateLimitGovernor {
    /**
     * @param {number|null} [maxRequestsPerMinute] Max requests allowed per 60-second window.
     * @param {number|null} [maxTokensPerMinute] Max tokens allowed per 60-second window.
     */
    constructor(maxRequestsPerMinute = null, maxTokensPerMinute = null) {
        this.maxRpm = maxRequestsPerMinute;
        this.maxTpm = maxTokensPerMinute;
        
        /** @type {Array<{timestamp: number, tokens: number}>} */
        this.history = [];
    }

    /**
     * Throttle execution by sleeping if sending the request would exceed configured limits.
     * 
     * @param {number} [estimatedTokens] Estimated count of tokens.
     * @returns {Promise<void>}
     */
    async throttleIfNeeded(estimatedTokens = 500) {
        if (!this.maxRpm && !this.maxTpm) {
            return;
        }

        while (true) {
            const currentTime = Date.now() / 1000;
            this._cleanOldRecords(currentTime);

            // Check if request count exceeds RPM threshold
            const rpmExceeded = this.maxRpm !== null && this.history.length >= this.maxRpm;

            // Check if token count exceeds TPM threshold
            const currentTokens = this.history.reduce((sum, item) => sum + item.tokens, 0);
            const tpmExceeded = this.maxTpm !== null && (currentTokens + estimatedTokens > this.maxTpm);

            if (!rpmExceeded && !tpmExceeded) {
                break;
            }

            // Calculate wait time until the oldest request falls out of the sliding window
            if (this.history.length > 0) {
                const oldestTime = this.history[0].timestamp;
                const sleepTime = (oldestTime + 60.0) - currentTime;
                if (sleepTime > 0) {
                    console.log(
                        `[EveryAI Governor] Approaching rate limits. ` +
                        `Throttling request, sleeping for ${sleepTime.toFixed(2)} seconds...`
                    );
                    await sleep(sleepTime * 1000);
                }
            } else {
                // Safeguard sleep to prevent CPU locks
                await sleep(500);
                break;
            }
        }
    }

    /**
     * Log a successful request in the governor's sliding window history.
     * 
     * @param {number} tokensUsed Exact count of prompt + completion tokens consumed.
     */
    recordRequest(tokensUsed) {
        if (!this.maxRpm && !this.maxTpm) {
            return;
        }
        this.history.push({
            timestamp: Date.now() / 1000,
            tokens: tokensUsed
        });
    }

    /**
     * Purge sliding-window requests older than 60 seconds.
     * 
     * @param {number} currentTime Current timestamp in seconds.
     * @private
     */
    _cleanOldRecords(currentTime) {
        const cutoff = currentTime - 60.0;
        while (this.history.length > 0 && this.history[0].timestamp <= cutoff) {
            this.history.shift();
        }
    }
}

File: py_package/everyai_core/governor.py
"""Local rate limit governor for everyai-core.

Tracks API requests and token counts locally inside a sliding window,
automatically throttling outgoing calls to avoid provider rate limit blocks.
"""

import time
from collections import deque
from typing import Any


class RateLimitGovernor:
    """Locally tracks request history to govern and throttle API request rates."""

    def __init__(self, max_requests_per_minute: int | None = None, max_tokens_per_minute: int | None = None):
        """Initialize the rate governor.

        Args:
            max_requests_per_minute: Maximum requests allowed per 60-second window.
            max_tokens_per_minute: Maximum tokens allowed per 60-second window.
        """
        self.max_rpm = max_requests_per_minute
        self.max_tpm = max_tokens_per_minute
        
        # Double-ended queue holding tuples of (timestamp, tokens_consumed)
        self.history: deque[tuple[float, int]] = deque()

    def throttle_if_needed(self, estimated_tokens: int = 500) -> None:
        """Throttle execution by sleeping if sending the request would exceed configured limits."""
        if not self.max_rpm and not self.max_tpm:
            return

        while True:
            current_time = time.time()
            self._clean_old_records(current_time)

            # Check if request count exceeds RPM threshold
            rpm_exceeded = self.max_rpm is not None and (len(self.history) >= self.max_rpm)

            # Check if token count exceeds TPM threshold
            current_tokens = sum(tokens for _, tokens in self.history)
            tpm_exceeded = self.max_tpm is not None and (current_tokens + estimated_tokens > self.max_tpm)

            if not rpm_exceeded and not tpm_exceeded:
                break

            # Calculate wait time until the oldest request falls out of the sliding window
            if self.history:
                oldest_time, _ = self.history[0]
                sleep_time = (oldest_time + 60.0) - current_time
                if sleep_time > 0:
                    # Clean ASCII printing to prevent UnicodeEncodeErrors on Windows terminals
                    print(
                        f"[EveryAI Governor] Approaching rate limits. "
                        f"Throttling request, sleeping for {sleep_time:.2f} seconds..."
                    )
                    time.sleep(sleep_time)
            else:
                # Safeguard sleep to prevent CPU locks
                time.sleep(0.5)
                break

    def record_request(self, tokens_used: int) -> None:
        """Log a successful request in the governor's sliding window history.

        Args:
            tokens_used: Exact count of prompt + completion tokens consumed by the call.
        """
        if not self.max_rpm and not self.max_tpm:
            return
        self.history.append((time.time(), tokens_used))

    def _clean_old_records(self, current_time: float) -> None:
        """Purge sliding-window requests older than 60 seconds."""
        cutoff = current_time - 60.0
        while self.history and self.history[0][0] <= cutoff:
            self.history.popleft()
